feat: add C implementation for `@stdlib/stats/base/dists/logistic/entropy`

DhruvArvindSingh opened the pull request:
Resolves #3734 .

## Description

> What is the purpose of this pull request?

This pull request:

-   Adds C implementation for `@stdlib/stats/base/dists/logistic/entropy`

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves #3734 
-   progresses #3692   

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

DhruvArvindSingh committed:
feat: add c implementation for @stdlib/stats/base/dists/logistic/entropy

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: passed
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

Planeshifter commented:
<!-- stdlib-bot-tracking-issue-closure-check -->
:warning: **Tracking Issue Closure Warning** :warning:

I noticed your PR description contains closing keywords ("Resolves", "Closes", or "Fixes") referencing a "Tracking Issue".

**Why this matters:**
Tracking issues should typically remain open until all related sub-issues are completed. GitHub automatically closes issues with such closing keywords when the PR is merged.

**Required action:**
Use "Progresses:" or "Ref:" instead to reference the tracking issue without automatically closing it.

Thank you for your contribution to the project!

Planeshifter committed:
Merge branch 'develop' into develop_#3734

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/logistic/entropy/README.md:
````diff
@@ -131,6 +131,89 @@ for ( i = 0; i < 10; i++ ) {
+double stdlib_base_dists_logistic_entropy( const double mu, const double s );
+```
+
+</section>
````
Missing newlines here and below between section tags, headings, and the like.

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/logistic/entropy/benchmark/benchmark.native.js:
```diff
@@ -0,0 +1,71 @@
+	mu = new Float64Array( len );
+	s = new Float64Array( len );
+	for ( i = 0; i < len; i++ ) {
+		mu[ i ] = ( randu() * 100.0 ) - 50.0;
```
It would be good to update this to draw the random numbers directly from the desired `uniform` distribution.

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/logistic/entropy/src/main.c:
```diff
@@ -0,0 +1,59 @@
+/**
+* Returns the differential entropy for a logistic distribution with location `mu` and scale `s`.
+*
+* @param {number} mu - location parameter
```
JSDoc here should be replaced with C Doxygen-style comments.

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/logistic/entropy/lib/native.js:
```diff
@@ -0,0 +1,62 @@
+
+/**
+* Returns the differential entropy for a logistic distribution with location `mu` and scale `s`.
+*
+* @param {number} mu - location parameter
```
```suggestion
*
* @private
* @param {number} mu - location parameter
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/logistic/entropy/benchmark/c/benchmark.c:
```diff
@@ -0,0 +1,141 @@
+	int i;
+
+	for ( i = 0; i < 100; i++ ) {
+		mu[ i ] = random_uniform( 0.0, 100.0 ) - 50.0;
+		s[ i ] = random_uniform( 0.0, 20.0 ) + STDLIB_CONSTANT_FLOAT64_EPS;
```
```suggestion
		mu[ i ] = random_uniform( -50.0, 50.0 );
		s[ i ] = random_uniform( STDLIB_CONSTANT_FLOAT64_EPS, 20.0 );
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/logistic/entropy/README.md:
```diff
@@ -131,6 +131,89 @@ for ( i = 0; i < 10; i++ ) {
+        mu = random_uniform( 0.0, 10.0 ) - 5.0;
+        s = random_uniform( 0.0, 20.0 );
+        y = stdlib_base_dists_logistic_entropy( mu, s );
+        printf( "µ: %lf, s: %lf, Ent(X;µ,s): %lf\n", mu, s, y );
```
```suggestion
        printf( "µ: %lf, s: %lf, h(X;µ,s): %lf\n", mu, s, y );
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/logistic/entropy/examples/c/example.c:
```diff
@@ -0,0 +1,40 @@
+		mu = random_uniform( 0.0, 10.0 ) - 5.0;
+		s = random_uniform( 0.0, 20.0 );
+		y = stdlib_base_dists_logistic_entropy( mu, s );
+		printf( "µ: %lf, s: %lf, Ent(X;µ,s): %lf\n", mu, s, y );
```
```suggestion
		printf( "µ: %lf, s: %lf, h(X;µ,s): %lf\n", mu, s, y );
```

Planeshifter reviewed (changes requested):
Thanks for the PR!

Flagged a bunch of issues that will have to be addressed before it can move forward. Similar changes are likely need in other PRs of yours such as https://github.com/stdlib-js/stdlib/pull/4391, too.

Planeshifter commented:
/stdlib update-copyright-years

stdlib-bot committed:
chore: update copyright years

Planeshifter committed:
chore: address PR review feedback

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: passed
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/logistic/entropy/README.md:
```diff
@@ -131,6 +131,101 @@ for ( i = 0; i < 10; i++ ) {
+    int i;
+
+    for ( i = 0; i < 25; i++ ) {
+        mu = random_uniform( 0.0, 10.0 ) - 5.0;
```
```suggestion
        mu = random_uniform( -5.0, 5.0 );
```

Planeshifter committed:
chore: directly draw from desired distribution

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter reviewed (approved):
Let's get it in!
